### PR TITLE
Add metadata change support for processors

### LINKED ISSUE #25425: Meta fields are not handled consistently in different processors
**TLDR;** All processors should use `event.PutValue` when setting fields to have a consistent behaviour.

Some fields set in events are intended to provide information to the pipelines, for example `@timestamp` to set the timestamp of the event or `@metadata._id` to set its id. This is not being handled in a consistent way between processors.

For example when the id is set using the fingerprint processor, it works as expected:
```
  - fingerprint:
      fields: ["id"]
      target_field: "@metadata._id"
```
It doesn't set `@metadata._id` in the fields of the event, but it uses it as the id of the generated document.

On the other hand, if `add_fields` is used to set one of these special fields, the field is set in the event and not handled by the pipeline:
```
  - add_fields:
      target: "@metadata"
      fields:
        op_type: "index"
```
In this case the `index` bulk operation should be used, but it isn't, and `@metadata.op_type` is set as a field in the event.

The cause of this difference is that some processors use the [`event.PutValue` method](https://github.com/elastic/beats/blob/aacf9ecd9c494aa0908f61fbca82c906b16562a8/libbeat/beat/event.go#L80), that handles these cases, and other processors use lower level mechanisms as `event.Fields.Put`, or directly `events.Fields[key] = value`.

All processors should use `event.PutValue` when setting fields to have a consistent behaviour.

For confirmed bugs, please report:
- Version: All versions at least till 7.12.
- Steps to Reproduce: The following configuration should create index bulk requests so documents are updated when they have the same fingerprint, but instead `create` bulk requests are used and then the documents are not updated:
```
processors:
  - fingerprint:
      fields: ["id"]
      target_field: "@metadata._id"
  - add_fields:
      target: "@metadata"
      fields:
        op_type: "index"
```

As a workaround, expected result can be achieved setting `@metadata.op_type` with the `convert` processor, that uses `PutValue`:
```
processors:
  - fingerprint:
      fields: ["id"]
      target_field: "@metadata._id"
  - add_fields:
      target: "_tmp"
      fields:
        op_type: "index"
  - convert:
      fields:
      - from: "_tmp.op_type"
        to: "@metadata.op_type"
        type: "string"
  - drop_fields.fields:
    - "_tmp"
```

Related:
- Add docs about some of these `@metadata` fields https://github.com/elastic/beats/issues/23091

### PR DESCRIPTION
**Pull request by rdner**
## What does this PR do?

Some of the processors that are configured with a target field didn't
support special cases such as when the target field is `@timestamp` or a field in `@metadata`. This change is to make it consistent across all processors or to document why a particular processor does not need this change.

This PR continues the work done in https://github.com/elastic/beats/pull/30092

## Why is it important?

Since some processors (e.g. `fingerprint`) support setting `@metadata.*` fields the users can expect the same behaviour from the rest of the processors too. To avoid this confusion we must either support the same behaviour across all the processors or document clearly when it's not supported.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
~~- [ ] I have made corresponding change to the default configuration files~~
~~- [ ] I have added tests that prove my fix is effective or that my feature works~~ (existing tests pass, the existing behaviour should not change)
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

### Are not supposed to access `Meta` (have hard-coded field names):

* `./add_docker_metadata/add_docker_metadata.go` (sets `container.*` fields only)
* `./add_locale/add_locale.go` (the doc states "The processor adds the a `event.timezone` value to each event.")
* `./add_kubernetes_metadata/kubernetes.go` (sets `kubernetes.*` fields only)
* `./add_host_metadata/add_host_metadata.go` (sets `host.*` fields only)
* `./add_observer_metadata/add_observer_metadata.go` (sets `obverver.*` fields only)
* `./actions/include_fields.go` (the doc states "The `@timestamp` and type fields are always exported, even if they are not defined in the include_fields list.", so should be expected behaviour)
* `./actions/add_tags.go` (@kvch and I agreed that tags should be set only in fields and not metadata)
* `./actions/add_labels.go` (sets the `labels` field)
* `./add_cloud_metadata/add_cloud_metadata.go` (sets only cloud.* fields, not configurable to set `@metadata`)

### Do not modify anything:

* `./actions/drop_event.go`
* `./ratelimit/rate_limit.go`

### Use event-specific functions (support metadata change):
* `./actions/drop_fields.go` 
* `./actions/decode_base64_field.go`
* `./actions/decode_json_fields.go`
* `./actions/extract_field.go`
* `./actions/add_network_direction.go`
* `./actions/decompress_gzip_field.go`
* `./timestamp/timestamp.go`
* `./dns/dns.go`
* `./fingerprint/fingerprint.go`
* `./translate_sid/translatesid.go`
* `./add_id/add_id.go`
* `./urldecode/urldecode.go`
* `./registered_domain/registered_domain.go`
* `./decode_xml/decode_xml.go`
* `./convert/convert.go`
* `./communityid/communityid.go`
* `./extract_array/extract_array.go`
* `./decode_csv_fields/decode_csv_fields.go`
* `./dissect/processor.go`
* `./decode_xml_wineventlog/processor.go`
* `./script/processor.go`
* **`./actions/add_fields.go` [fixed in #30092]**
* **`./add_process_metadata/add_process_metadata.go` [fixed in this PR]**
* **`./actions/detect_mime_type.go` [fixed in this PR]**
* **`./actions/truncate_fields.go` [fixed in this PR]**
* **`./actions/replace.go` [fixed in this PR]**
* **`./actions/rename.go` [fixed in this PR]**
* **`./actions/copy_fields.go` [fixed in this PR]**

## Related issues

- Relates #30092
- Closes #25425

### CODE REVIEW

**Comment by cmacknz** [comment]
The changes themselves look good to me. The one concern I have is that we have no tests ensuring this inconsistency with the metadata fields won't happen again. Could we add a test for each affected processor to ensure that using `@metadata` actually modifies the metadata of the event and doesn't just add a new field?

The original reason this happened is that it isn't obvious at all from the code that the treatment of `@metadata` is different between the different types/methods for handling events. I would prefer if we can do something to enforce the correct behaviour in the future.

**Review by cmacknz — CHANGES_REQUESTED** [review]
> The changes themselves look good to me. The one concern I have is that we have no tests ensuring this inconsistency with the metadata fields won't happen again. Could we add a test for each affected processor to ensure that using @metadata actually modifies the metadata of the event and doesn't just add a new field?
> 
> The original reason this happened is that it isn't obvious at all from the code that the treatment of @metadata is different between the different types/methods for handling events. I would prefer if we can do something to enforce the correct behaviour in the future.

Should have requested changes with this comment.

File: libbeat/processors/add_docker_metadata/add_docker_metadata.go [hunk 1]
```diff
@@ -132,6 +132,8 @@ func lazyCgroupCacheInit(d *addDockerMetadata) {
 	}
 }
 
+// Run runs the processor that adds container-related fields to the event
+// This processor does not modify any `Meta` fields in the event, it accesses and modifies `event.Fields` directly
```
Comment (kvch) [inline libbeat/processors/add_docker_metadata/add_docker_metadata.go:136]: Why did you change it? Do we win anything by touching processors that are not impacted?

Comment (rdner) [reply libbeat/processors/add_docker_metadata/add_docker_metadata.go:136]: yes, now it looks more obvious that the processor is not allowed to edit metadata and works with the `Fields` only. I still think it's a good idea.

File: libbeat/processors/add_locale/add_locale.go [hunk 2]
```diff
@@ -89,7 +89,7 @@ func New(c *common.Config) (processors.Processor, error) {
 func (l addLocale) Run(event *beat.Event) (*beat.Event, error) {
 	zone, offset := time.Now().Zone()
 	format := l.Format(zone, offset)
-	event.PutValue("event.timezone", format)
+	event.Fields.Put("event.timezone", format)
```
Comment (kvch) [inline libbeat/processors/add_locale/add_locale.go:92]: TBH I prefer that in general no one updates the` Fields` attribute directly. I especially would not put extra effort into changing processors that are not impacted by the bug.

**Comment by rdner** [comment]
@cmacknz since it's quite a bit addition, I think it's a good idea to continue working on tests in a separate PR, what do you think?

File: libbeat/processors/add_locale/add_locale.go (hunk 2 above)
Comment (rdner) [reply libbeat/processors/add_locale/add_locale.go:92]: Well, if there is a strong opinion against it – I can rollback the change. But I still think it's way easier to navigate the code this way. For example, when you find usages of `PutValue` you find all the occurrences where it's meant to be used with metadata support, all the occurrences of `Fields.Put` indicate no metadata support.

After reading all the 39 processors and looking for this distinction I find this useful :)

**Comment by cmacknz** [comment]
> @cmacknz since it's quite a bit addition, I think it's a good idea to continue working on tests in a separate PR, what do you think?

I'm not opposed to it, but I generally prefer when PRs come with tests proving they fixed what they say they did :) The existing tests tell us none of the existing functionality is broken, so in theory merging this won't break anything that isn't already broken (metadata field handling).

**Comment by rdner** [comment]
@cmacknz I totally agree, it's just to do it right I need to introduce 24 new tests for each processor that is **supposed** to support changing metadata. I sure can continue in this PR, most likely will finish the tests tomorrow.

File: libbeat/processors/add_locale/add_locale.go (hunk 2 above)
Comment (kvch) [reply libbeat/processors/add_locale/add_locale.go:92]: I prefer to have only one way of extending events with new fields. `PutValue` is safe for both fields and metadata. Please roll back these changes.

> After reading all the 39 processors and looking for this distinction I find this useful :)

This way other developers do not have to read all of the 39 processors to find this pattern. :) We should simplify processor development so there are fewer possibilities for error. Maybe it is also worth adding a few sentences to our developer guide (if we have any about processor development).

**Review by kvch — CHANGES_REQUESTED** [review]
Please do not change existing `event.PutValue` to `event.Fields.Put`.

**Comment by rdner** [comment]
@cmacknz I've added tests for the processors that support metadata changes:

* `./actions/drop_fields.go` 
* `./actions/decode_base64_field.go`
* `./actions/decode_json_fields.go`
* `./actions/extract_field.go`
* `./actions/add_network_direction.go`
* `./actions/decompress_gzip_field.go`
* `./timestamp/timestamp.go`
* `./dns/dns.go`
* `./fingerprint/fingerprint.go`
* `./translate_sid/translatesid.go`
* `./add_id/add_id.go`
* `./urldecode/urldecode.go`
* `./registered_domain/registered_domain.go`
* `./decode_xml/decode_xml.go`
* `./convert/convert.go`
* `./communityid/communityid.go`
* `./extract_array/extract_array.go`
* `./decode_csv_fields/decode_csv_fields.go`
* `./dissect/processor.go`
* `./decode_xml_wineventlog/processor.go`
* ~~`./script/processor.go`~~ (tests were already there https://github.com/elastic/beats/blob/8c79b67906ab4399dd9e575e5f2c4005498c32f3/libbeat/processors/script/javascript/beatevent_v0_test.go#L115-L127)
* ~~`./actions/add_fields.go`~~ (tests were added in #30092)
* `./add_process_metadata/add_process_metadata.go`
* `./actions/detect_mime_type.go`
* `./actions/truncate_fields.go`
* `./actions/replace.go`
* `./actions/rename.go`
* `./actions/copy_fields.go`


@kvch I removed the `event.PutValue` to `event.Fields.Put` changes.

**Review by cmacknz — APPROVED** [review]
Thanks, that's a lot of tests :)

**Review by kvch — APPROVED** [review]
